cloud-controller-manager/do/firewall_controller.go

```diff
@@ -372,3 +409,38 @@ func (fc *firewallCache) updateCache(currentFirewall *godo.Firewall) {
 	defer fc.mu.Unlock()
 	fc.firewall = currentFirewall
 }
+
+func (fm *firewallManager) incrementPrometheusCounter(resp *godo.Response, absentResponse *absentResponse) {
+	var labels prometheus.Labels
+	if absentResponse.expected {
+		labels = prometheus.Labels{"http_code": "", "http_method": "", "success": strconv.FormatBool(absentResponse.isSuccess)}
+	} else if resp == nil {
+		labels = prometheus.Labels{"http_code": "", "http_method": "", "success": "false"}
+	} else if resp != nil && resp.Request != nil {
+		labels = prometheus.Labels{
+			"http_code":   strconv.FormatInt(int64(resp.StatusCode), 10),
+			"http_method": resp.Request.Method,
+			"success":     strconv.FormatBool((resp.StatusCode >= 200 && resp.StatusCode < 300)),
+		}
+	} else {
+		labels = prometheus.Labels{"http_code": "", "http_method": "", "success": "false"}
+	}
+
+	fm.metrics.apiRequestCounter.With(labels).Inc()
+}
+
+func (fm *firewallManager) observeDuration(resp *godo.Response) {
+	var labels prometheus.Labels
+	if resp != nil && resp.Request != nil {
+		labels = prometheus.Labels{
+			"http_code":   strconv.FormatInt(int64(resp.StatusCode), 10),
+			"http_method": resp.Request.Method,
+			"success":     strconv.FormatBool((resp.StatusCode >= 200 && resp.StatusCode < 300)),
+		}
+	} else {
+		labels = prometheus.Labels{"http_code": "", "http_method": "", "success": "false"}
+	}
+	fm.incrementPrometheusCounter(resp, &absentResponse{expected: false})
+	t := prometheus.NewTimer(fm.metrics.apiRequestDuration.With(labels))
```

Comment:
Nice catch. Because of the need to start the timer before the operation starts and the way I'm deriving labels from the response (which comes from performing the operation in order to get the response to pass into a helper function):
```go
func (fm *firewallManager) observeDuration(resp *godo.Response) {
	var labels prometheus.Labels
	if resp != nil && resp.Request != nil {
		labels = prometheus.Labels{
			"http_code":   strconv.FormatInt(int64(resp.StatusCode), 10),
			"http_method": resp.Request.Method,
			"success":     strconv.FormatBool((resp.StatusCode >= 200 && resp.StatusCode < 300)),
		}
	} else {
		labels = prometheus.Labels{"http_code": "", "http_method": "", "success": "false"}
	}
	fm.incrementPrometheusCounter(resp, &absentResponse{expected: false})
	t := prometheus.NewTimer(fm.metrics.apiRequestDuration.With(labels))
	t.ObserveDuration()
}
```

Because of this, it's a little trickier than just starting the timer before the operation when we get labels based off of the operation. I'll have to refactor the way I make labels.

Reply:
@timoreimann ^